src/storage/concurrency_manager/memory_lock.rs

```diff
@@ -0,0 +1,275 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::lock_table::OrderedLockMap;
+
+use event_listener::Event;
+use kvproto::kvrpcpb::LockInfo;
+use parking_lot::Mutex;
+use std::{
+    future::Future,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicBool, AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+const INIT_REF_COUNT: usize = usize::MAX;
+
+pub struct MemoryLock {
+    key: Vec<u8>,
+    mutex_locked: AtomicBool,
+    mutex_event: Event,
+    lock_info: Mutex<Option<LockInfo>>,
+    pessimistic_event: Event,
+    ref_count: AtomicUsize,
```

Comment:
Why do we need an internal ref count?

Reply:
I tried but found it really hard to use `Arc`'s strong count to do the same thing. The problem is that we have a special logic that if `lock_info` is `Some`, `MemoryLock` can't be freed and removed from the map.

Reply:
Could you make `lock_info: Mutex<Option<(Arc<MemoryLock>, LockInfo)>>`?

Reply:
Actually we'll have another small problem if we depend on the destructor of `MemoryLock` to remove it from the map: we won't be able to use a trait for a the map.
Because we need to access the map in `MemoryLock::drop`, we need `MemoryLock<M: Map>`. Then the map itself will be `BTreeMap<Vec<u8>, MemoryLock<Self>>`, so we can a cyclic type.